x/shield/keeper/withdraw.go

```diff
@@ -105,3 +130,218 @@ func (k Keeper) DequeueCompletedWithdrawQueue(ctx sdk.Context) {
+		timeslice := []types.Withdraw{}
+		value := withdrawTimesliceIterator.Value()
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(value, &timeslice)
+
+		for _, withdraw := range timeslice {
+			if withdraw.Address.Equals(provider) {
+				withdraws = append(withdraws, withdraw)
+			}
+		}
+	}
+
+	// Delay withdraws, starting with the candidates
+	// with the oldest withdraw completion time.
+	remaining := amount
+	for i := len(withdraws) - 1; i >= 0; i-- {
+		if !remaining.IsPositive() {
+			break
+		}
+		// Remove from withdraw queue.
+		timeSlice := k.GetWithdrawQueueTimeSlice(ctx, withdraws[i].CompletionTime)
+		if len(timeSlice) > 1 {
+			for j := 0; j < len(timeSlice); j++ {
+				if timeSlice[j].Address.Equals(provider) {
+					timeSlice = append(timeSlice[:j], timeSlice[j+1:]...)
+					k.SetWithdrawQueueTimeSlice(ctx, withdraws[i].CompletionTime, timeSlice)
+					break
+				}
+			}
+		} else {
+			k.RemoveTimeSliceFromWithdrawQueue(ctx, withdraws[i].CompletionTime)
+		}
+
+		// Adjust the withdraw completion time and re-insert.
+		withdraws[i].CompletionTime = delayedTime
+		k.InsertWithdrawQueue(ctx, withdraws[i])
+
+		remaining = remaining.Sub(withdraws[i].Amount)
+	}
+
+	if remaining.IsPositive() {
+		panic("failed to delay enough withdraws")
+	}
+
+	return nil
+}
+
+func (k Keeper) DelayUnbonding(ctx sdk.Context, provider sdk.AccAddress, amount sdk.Int, delay time.Duration) error {
+	// Retrieve delay candidates, which are unbondings
+	// ending before the delay duration from now.
+	delayedTime := ctx.BlockTime().Add(delay)
+	ubds := k.getUnbondingsByProviderMaturingByTime(ctx, provider, delayedTime)
+
+	// Delay unbondings, starting with the candidates
+	// with the oldest unbonding completion time.
+	remaining := amount
+	for i := len(ubds) - 1; i >= 0; i-- {
+		if !remaining.IsPositive() {
+			break
+		}
+		// Remove from unbonding queue.
+		timeSlice := k.sk.GetUBDQueueTimeSlice(ctx, ubds[i].CompletionTime)
+		if len(timeSlice) > 1 {
+			for j := 0; j < len(timeSlice); j++ {
+				if timeSlice[j].DelegatorAddress.Equals(provider) && timeSlice[j].ValidatorAddress.Equals(ubds[i].Validator) {
+					timeSlice = append(timeSlice[:j], timeSlice[j+1:]...)
+					k.sk.SetUBDQueueTimeSlice(ctx, ubds[i].CompletionTime, timeSlice)
+					break
+				}
+			}
+		} else {
+			k.sk.RemoveUBDQueue(ctx, ubds[i].CompletionTime)
+		}
+
+		unbondingDels, found := k.sk.GetUnbondingDelegation(ctx, provider, ubds[i].Validator)
+		if !found {
+			panic("unbonding list was not found for the given provider-validator pair")
+		}
+
+		found = false
+		amount := sdk.ZeroInt()
+		for j := 0; j < len(unbondingDels.Entries); j++ {
+			if !found && unbondingDels.Entries[j].CompletionTime.Equal(ubds[i].CompletionTime) {
+				unbondingDels.Entries[j].CompletionTime = delayedTime
+				found = true
+				amount = unbondingDels.Entries[j].Balance
+			} else if found && unbondingDels.Entries[j].CompletionTime.Before(unbondingDels.Entries[j-1].CompletionTime) {
```

Comment:
remove else and remove found checking

if !... {
   break
}
...

or

if ... {
  ....
  continue
}
break